packages/edit-post/src/components/header/index.js

```diff
@@ -82,22 +71,21 @@ function Header( { setEntitiesSavedStatesCallback, initialPost } ) {
 			isPublishSidebarOpened:
 				select( editorStore ).isPublishSidebarOpened(),
 			showIconLabels: getPreference( 'core', 'showIconLabels' ),
+			hasFixedToolbar: getPreference( 'core', 'fixedToolbar' ),
 		};
 	}, [] );
 
-	const { showFixedToolbar } = useShowBlockTools();
-	const showTopToolbar = isLargeViewport && showFixedToolbar;
+	const hasTopToolbar = isLargeViewport && hasFixedToolbar;
 
 	const [ isBlockToolsCollapsed, setIsBlockToolsCollapsed ] =
 		useState( true );
-	const hasBlockSelection = !! blockSelectionStart;
 
-	useEffect( () => {
-		// If we have a new block selection, show the block tools
-		if ( blockSelectionStart ) {
-			setIsBlockToolsCollapsed( false );
-		}
-	}, [ blockSelectionStart ] );
```

Comment:
This `useEffect` was causing a rerender each time a new block was selected. It's only necessary when top toolbar is on, but the state is necessary when the document bar is showing (when editing a post template), so this has been changed to a function passed to the ContextualToolbar component.